### PR TITLE
Issue/7531 race condition in exporter file upload

### PR DESCRIPTION
**Pull request by Hugo-Inmanta**
# Description

- [x] Make the file upload endpoint idempotent: if the file already exists, silently ignore the request
- [x] Double check that the exporter can handle this new behavior

closes https://github.com/inmanta/inmanta-core/issues/7531

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
~~- [ ] Correct, in line with design~~
- [x] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )
~~- [ ] If this PR fixes a race condition in the test suite, also push the fix to the relevant stable branche(s) (see [test-fixes](https://internal.inmanta.com/development/core/tasks/build-master.html#test-fixes) for more info)~~

### CODE REVIEW

**Comment by Hugo-Inmanta** [comment]
> Double check that the exporter can handle this new behavior

For this point this seems ok as far as I can tell, and the test suite didn't go up in smoke. I'm not sure if this is enough or what else needs to be done to check this.

File: tests/protocol/test_protocol.py [hunk 1]
```diff
@@ -136,6 +136,28 @@ async def test_client_files_stat(client):
...
+        assert len(result.result["files"]) == 0
+
+    (hash, content, body) = make_random_file()
+
+    result = await client.stat_files(files=[hash])
+    assert result.code == 200
+    assert len(result.result["files"]) == 1
+
+    await upload_and_stat(body, client, hash)
+    await upload_and_stat(body, client, hash)
```
Comment (arnaudsjs) [inline tests/protocol/test_protocol.py:158]: To test whether the concurrency issue is resolved, you should run these two API calls concurrently.
```suggestion
    await asyncio.gather(upload_and_stat(body, client, hash),  upload_and_stat(body, client, hash))
```

File: changelogs/unreleased/7531-race-condition-in-exporter-file-upload.yml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+description: >
+  The upload_file endpoint will now silently ignore attempts to upload a file if a file with the same
+  hash was previously uploaded.
+issue-nr: 7531
+change-type: minor
+destination-branches: [master, iso7]
+sections:
+  minor-improvement: "{{description}}"
```
Comment (arnaudsjs) [inline changelogs/unreleased/7531-race-condition-in-exporter-file-upload.yml:8]: To me this change is just a bugfix. I don't really see why it needs a minor-improvement entry.

File: changelogs/unreleased/7531-race-condition-in-exporter-file-upload.yml [hunk 3]
```diff
@@ -0,0 +1,10 @@
+description: >
+  The upload_file endpoint will now silently ignore attempts to upload a file if a file with the same
+  hash was previously uploaded.
+issue-nr: 7531
+change-type: minor
+destination-branches: [master, iso7]
```
Comment (arnaudsjs) [inline changelogs/unreleased/7531-race-condition-in-exporter-file-upload.yml:6]: Is there a specific reason why we don't want this fix into iso6?

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: changelogs/unreleased/7531-race-condition-in-exporter-file-upload.yml (hunk 3 above)
Comment (Hugo-Inmanta) [reply changelogs/unreleased/7531-race-condition-in-exporter-file-upload.yml]: Hmm no actually. It felt more like a minor than a patch to me, but looking back again at it, this shouldn't drive to which branch it gets merged. I'll add iso6 and I'll try to be more careful about this in the future.

**Review by arnaudsjs — APPROVED** [review]

File: changelogs/unreleased/7531-race-condition-in-exporter-file-upload.yml [hunk 4]
```diff
@@ -0,0 +1,10 @@
+description: >
+  The upload_file endpoint will now silently ignore attempts to upload a file if a file with the same
+  hash was previously uploaded.
+issue-nr: 7531
+change-type: patch
+destination-branches: [master, iso7, iso6]
+sections:
+  bugfix: >
+    Fix race condition where exporting a file might fail if a file with similar content
```
Comment (sanderr) [inline changelogs/unreleased/7531-race-condition-in-exporter-file-upload.yml:9]:
```suggestion
    Fix race condition where exporting a file might fail if a file the same content
```

File: src/inmanta/protocol/methods.py [hunk 5]
```diff
@@ -876,7 +876,7 @@ def upload_code_batched(tid: uuid.UUID, id: int, resources: dict):
     :param tid: The id of the environment to which the code belongs.
     :param id: The version number of the configuration model.
     :param resources: A dictionary where each key is a string representing a resource type.
-                  For each resource type, the value is a dictionary. This nested dictionary's keys are file names,
+                  For each resource type, the value is a dictionary. This nested dictionary's keys are file hashes,
```
Comment (sanderr) [inline src/inmanta/protocol/methods.py:879]: I think they're technically both, but perhaps this is an improvement.

**Review by sanderr — CHANGES_REQUESTED** [review]

File: src/inmanta/protocol/methods.py (hunk 5 above)
Comment (Hugo-Inmanta) [reply src/inmanta/protocol/methods.py:879]: To me `names` implies a clear text file name

Comment (sanderr) [reply src/inmanta/protocol/methods.py:879]: Well, it is. Iirc this is the name we use to store this on the file system. But perhaps I don't remember correctly.

**Review by sanderr — APPROVED** [review]

File: src/inmanta/protocol/methods.py (hunk 5 above)
Comment (Hugo-Inmanta) [reply src/inmanta/protocol/methods.py:879]: Yes I think these keys are hashes: we use the hash to call into upload_code_batched:  https://github.com/inmanta/inmanta-core/blob/a6b57a5ba24d9bccbee9884482de0510a5ab153e/src/inmanta/export.py#L107-L112

**Pull request closed without merging**